akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -0,0 +1,211 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import java.io.{ BufferedWriter, File, FileWriter, Writer }
+import java.util.concurrent.Executors
+
+import akka.Done
+import akka.stream.ActorAttributes._
+import akka.stream.Supervision._
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.Utils._
+import akka.stream.testkit.scaladsl.TestSource
+import org.scalatest.BeforeAndAfter
+
+import scala.concurrent.{ Await, ExecutionContext, Future, Promise }
+
+class FoldResourceAsyncSinkSpec extends StreamSpec(UnboundedMailboxConfig) with BeforeAndAfter {
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+
+  val ioExecutor = Executors.newFixedThreadPool(4)
+  val ioExecutionContext = ExecutionContext.fromExecutor(ioExecutor)
+
+  override def afterTermination(): Unit = {
+    ioExecutor.shutdown()
+  }
+
+  val manyLines = {
+    ("a" * 100 + "\n") * 10 +
+      ("b" * 100 + "\n") * 10 +
+      ("c" * 100 + "\n") * 10 +
+      ("d" * 100 + "\n") * 10 +
+      ("e" * 100 + "\n") * 10 +
+      ("f" * 100 + "\n") * 10
+  }
+  val manyLinesArray = manyLines.split("\n")
+
+  var sinkFile: File = _
```

Comment:
Ah, that is true, ignore that part of what I wrote :) 

You do not have to actually do anything resource related to test the sink though. You just need something mutable that you can write to and then make that available to the test case assertions in close. To verify that the callbacks where called you don't even have to do that, but can use promises or some other concurrency tool just like you already do.